File: 1030 - Matrix Cells in Distance Order.py
from collections import deque


class Solution:
    def allCellsDistOrder(self, rows: int, cols: int, rCenter: int, cCenter: int) -> List[List[int]]:
        res = [[rCenter, cCenter]]
        q = deque()
        visited = set()
        visited.add((rCenter, cCenter))
        q.append([rCenter, cCenter])
        n = rows + cols
        dirs = [[0, 1], [0, -1], [-1, 0], [1, 0]]
        while len(q) != 0:
            cr, cc = q.popleft()
            for dr, dc in dirs:
                r = cr + dr
                c = cc + dc
                if 0 <= r < rows and 0 <= c < cols and (r, c) not in visited:
                    res.append([r, c])
                    visited.add((r, c))
                    q.append([r, c])
        return res

File: 1030 - Matrix Cells in Distance Order.js
/**
 * @param {number} rows
 * @param {number} cols
 * @param {number} rCenter
 * @param {number} cCenter
 * @return {number[][]}
 */
var allCellsDistOrder = function (rows, cols, rCenter, cCenter) {
    let res = new Array(rows * cols);
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            res[i * cols + j] = [i, j];
        }
    }
    res.sort((a, b) => {
        let ad = Math.abs(a[0] - rCenter) + Math.abs(a[1] - cCenter);
        let bd = Math.abs(b[0] - rCenter) + Math.abs(b[1] - cCenter);
        return ad - bd;
    });
    return res;
};